rust-runtime/aws-smithy-http/src/header.rs

```diff
@@ -164,43 +164,106 @@ where
+        let next_delim = input.iter().position(|&b| b == b',').unwrap_or(input.len());
+        let (first, next) = input.split_at(next_delim);
+        let first = std::str::from_utf8(first)
+            .map_err(|_| ParseError::new_with_message("header was not valid utf8"))?;
+        Ok((Cow::Borrowed(first), then_delim(next).unwrap()))
+    }
+
+    /// Reads a header value that is surrounded by quotation marks and may have escaped
+    /// quotes inside of it.
+    fn read_quoted_value(input: &[u8]) -> Result<(Cow<'_, str>, &[u8]), ParseError> {
+        for index in 0..input.len() {
+            match input[index] {
+                b'"' if index == 0 || input[index - 1] != b'\\' => {
+                    let mut inner =
+                        Cow::Borrowed(std::str::from_utf8(&input[0..index]).map_err(|_| {
+                            ParseError::new_with_message("header was not valid utf8")
+                        })?);
+                    if inner.contains("\\\"") {
+                        inner = Cow::Owned(inner.replace("\\\"", "\""));
+                    }
+                    let rest = then_delim(&input[(index + 1)..])?;
+                    return Ok((inner, rest));
+                }
+                _ => {}
+            }
+        }
+        Err(ParseError::new_with_message(
+            "header value had quoted value without end quote",
+        ))
+    }
+
+    fn then_delim(s: &[u8]) -> Result<&[u8], ParseError> {
+        if s.is_empty() {
+            Ok(s)
+        } else if s.starts_with(b",") {
+            Ok(&s[1..])
+        } else {
+            Err(ParseError::new_with_message("expected delimiter `,`"))
+        }
+    }
+}
+
 /// Read one comma delimited value for `FromStr` types
 fn read_one<'a, T>(
     s: &'a [u8],
     f: &impl Fn(&str) -> Result<T, ParseError>,
 ) -> Result<(T, &'a [u8]), ParseError> {
-    let (head, rest) = split_at_delim(s);
-    let head = std::str::from_utf8(head)
-        .map_err(|_| ParseError::new_with_message("header was not valid utf8"))?;
-    Ok((f(head.trim())?, rest))
+    let (value, rest) = parse_multi_header::read_value(s)?;
+    Ok((f(value.trim())?, rest))
 }
 
-fn split_at_delim(s: &[u8]) -> (&[u8], &[u8]) {
-    let next_delim = s.iter().position(|b| b == &b',').unwrap_or(s.len());
-    let (first, next) = s.split_at(next_delim);
-    (first, then_delim(next).unwrap())
-}
-
-fn then_delim(s: &[u8]) -> Result<&[u8], ParseError> {
-    if s.is_empty() {
-        Ok(s)
-    } else if s.starts_with(b",") {
-        Ok(&s[1..])
+/// Conditionally quotes and escapes a header value if the header value contains a comma or quote.
+pub fn quote_header_value<'a>(value: impl Into<Cow<'a, str>>) -> Cow<'a, str> {
+    let value = value.into();
+    if value.contains('"') || value.contains(',') {
+        Cow::Owned(format!("\"{}\"", value.replace('"', "\\\"")))
     } else {
-        Err(ParseError::new_with_message("expected delimiter `,`"))
+        value
     }
 }
```

Comment:
looking at the spec, it seems like there are other characters that may need to be quoted?

Reply:
It looks like I missed `(`, and `)`. Will add.